pymc3/gp/gp.py

```diff
@@ -68,17 +125,231 @@ def _random(mean, cov, size=None):
+        return mean, cov
+
+    def logp(self, y):
+        mean = self.m(self.X)
+        L = cholesky(stabilize(self.K(self.X)) + self.Kn(self.X))
+        return MvNormal.dist(mu=mean, chol=L).logp(y)
 
-        return MvNormal.dist(mu, Sigma).logp(Y)
 
+class GPSparseConjugate(GPBase, Continuous):
+    """Sparse Gausian process approximation
 
-def sample_gp(trace, gp, X_values, samples=None, obs_noise=True, model=None, random_seed=None, progressbar=True,
-              chol_const=True):
+    Parameters
+    ----------
+    X : array
+        Array of predictor variables.
+    mean_func : Mean
+        Mean function of Gaussian process
+    cov_func : Covariance
+        Covariance function of Gaussian process
+    sigma2 : scalar or array
+        Observation variance
+    approx : string
+        Allowed values are 'FITC' and 'VFE'
+    inducing_points : array
+        Grid of points to evaluate Gaussian process over.
+    """
+    def __init__(self, X, mean_func, cov_func, sigma2, approx, inducing_points, *args, **kwargs):
+        self.X = X
+        self.nf = self.X.shape[0]
+        self.K = cov_func
+        self.m = mean_func
+        self.mean = self.mode = self.m(X)
+
+        self.sigma2 = sigma2
+        self.approx = approx
+        kwargs.setdefault("shape", X.squeeze().shape)
+
+        self.Xu = inducing_points
+        self.nu = self.Xu.shape[0]
+        super(GPSparseConjugate, self).__init__(*args, **kwargs)
+
+    def prior(self, obs_noise=False):
+        Kuu = self.K(self.Xu)
+        Kuf = self.K(self.Xu, self.X)
+        Luu = cholesky(stabilize(Kuu))
+        A = solve_lower(Luu, Kuf)
+        Kffd = self.K(self.X, diag=True)
+        Qff = tt.dot(tt.transpose(A), A)
+        mean = self.m(self.X)
+        if obs_noise:
+            cov = Qff - (tt.diag(Qff) - Kffd) + self.sigma2 * tt.eye(self.nf)
+        else:
+            cov = stabilize(Qff - (tt.diag(Qff) - Kffd))
+        return mean, cov
+
+    def conditional(self, Xs, y, obs_noise=False):
+        Kuu = self.K(self.Xu)
+        Kuf = self.K(self.Xu, self.X)
+        Luu = cholesky(stabilize(Kuu))
+        A = solve_lower(Luu, Kuf)
+        Qffd = tt.sum(A * A, 0)
+        if self.approx == "FITC":
+            Kffd = self.K(self.X, diag=True)
+            Lamd = tt.clip(Kffd - Qffd, 0.0, np.inf) + self.sigma2
+        elif self.approx == "VFE":
+            Lamd = tt.ones_like(Qffd) * self.sigma2
+        else:
+            raise ValueError("unknown approximation string", self.approx)
+        A_l = A / Lamd
+        L_B = cholesky(tt.eye(self.nu) + tt.dot(A_l, tt.transpose(A)))
+        r = y - self.m(self.X)
+        r_l = r / Lamd
+        c = solve_lower(L_B, tt.dot(A, r_l))
+        Kus = self.K(self.Xu, Xs)
+        As = solve_lower(Luu, Kus)
+        mean = tt.dot(tt.transpose(As), solve_upper(tt.transpose(L_B), c))
+        C = solve_lower(L_B, As)
+        if obs_noise:
+            cov = self.K(Xs, Xs) - tt.dot(tt.transpose(As), As)\
```

Comment:
Avoid continuation character throughout. Enclose selection in parentheses instead.